src/promptflow/promptflow/_core/openai_injector.py

```diff
@@ -81,22 +111,41 @@ def get_aoai_telemetry_headers() -> dict:
 
 
 def inject_operation_headers(f):
-    @functools.wraps(f)
-    def wrapper(*args, **kwargs):
-        # Inject headers from operation context, overwrite injected header with headers from kwargs.
-        injected_headers = get_aoai_telemetry_headers()
-        original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")
-        if original_headers and isinstance(original_headers, dict):
-            injected_headers.update(original_headers)
-        kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(extra_headers=injected_headers)
+    if asyncio.iscoroutinefunction(f):
+
+        @functools.wraps(f)
+        async def wrapper(*args, **kwargs):
+            # Inject headers from operation context, overwrite injected header with headers from kwargs.
+            injected_headers = get_aoai_telemetry_headers()
+            original_headers = kwargs.get("headers") if IS_LEGACY_OPENAI else kwargs.get("extra_headers")
+            if original_headers and isinstance(original_headers, dict):
+                injected_headers.update(original_headers)
+            kwargs.update(headers=injected_headers) if IS_LEGACY_OPENAI else kwargs.update(
+                extra_headers=injected_headers
+            )
```

Comment:
Let's extract this part into a subfunction.

Reply:
Yes. I'd like to do this.